### PR TITLE
WriteUnPrepared: Add support for recovering WriteUnprepared transactions

### PR DESCRIPTION
**Pull request by lth**
This adds support for recovering WriteUnprepared transactions through the following changes:
- The information in `RecoveredTransaction` is extended so that it can reference multiple batches.
- `MarkBeginPrepare` is extended with a bool indicating whether it is an unprepared begin, and this is passed down to `InsertRecoveredTransaction` to indicate whether the current transaction is prepared or not.
- `WriteUnpreparedTxnDB::Initialize` is overridden so that it will rollback unprepared transactions from the recovered transactions. This can be done without updating the prepare heap/commit map, because this is before the DB has finished initializing, and after writing the rollback batch, those data structures should not contain information about the rolled back transaction anyway.

Commit/Rollback of live transactions is still unimplemented and will come later.

### CODE REVIEW

File: utilities/transactions/write_unprepared_txn_db.cc [hunk 1]
```diff
@@ -15,6 +15,303 @@
...
+    WriteOptions w_options;
+    w_options.sync = true;
+    TransactionOptions t_options;
+
+    uint64_t first_log_number =
+        recovered_trx->batches_.begin()->second.log_number_;
+    Transaction* real_trx = BeginTransaction(w_options, t_options, nullptr);
+    assert(real_trx);
+    real_trx->SetLogNumber(first_log_number);
+    real_trx->SetId(0);
```
Comment (maysamyabandeh) [inline utilities/transactions/write_unprepared_txn_db.cc]: why 0?

File: utilities/transactions/write_unprepared_txn_db.cc [hunk 2]
```diff
@@ -15,6 +15,303 @@
...
+
+    s = real_trx->SetName(recovered_trx->name_);
+    if (!s.ok()) {
+      break;
+    }
+
+    auto wupt =
+        static_cast_with_check<WriteUnpreparedTxn, Transaction>(real_trx);
```
Comment (maysamyabandeh) [inline utilities/transactions/write_unprepared_txn_db.cc]: Will there be any real_trx->RebuildFromWriteBatch or equivalent to that?

File: utilities/transactions/write_unprepared_txn_db.cc [hunk 3]
```diff
@@ -15,6 +15,303 @@
 
 namespace rocksdb {
 
+Status WriteUnpreparedTxnDB::RollbackRecoveredTransaction(
```
Comment (maysamyabandeh) [inline utilities/transactions/write_unprepared_txn_db.cc]: Can you document the rationale for rolling back a RecoveredTransaction instead of reconstructing a Transaction and calling rollback on it?

File: utilities/transactions/write_unprepared_txn_db.cc [hunk 4]
```diff
@@ -15,6 +15,303 @@
...
+    const bool kDisableMemtable = true;
+    const size_t kOneBatch = 1;
+    uint64_t seq_used = kMaxSequenceNumber;
+    s = db_impl_->WriteImpl(w_options, &rollback_batch, nullptr, nullptr,
+                            kNoLogRef, !kDisableMemtable, &seq_used, kOneBatch);
+    if (!s.ok()) {
+      return s;
+    }
+
+    // If two_write_queues, then a second write must be done to release
```
Comment (maysamyabandeh) [inline utilities/transactions/write_unprepared_txn_db.cc]: Since we are in single threaded recovery mode, i do not think a write through the 2nd queue is necessary. We can simply publish the seq without being worried about concurrent updates.

File: utilities/transactions/write_unprepared_txn_db.cc [hunk 5]
```diff
@@ -15,6 +15,303 @@
 
 namespace rocksdb {
 
+Status WriteUnpreparedTxnDB::RollbackRecoveredTransaction(
+    const DBImpl::RecoveredTransaction* rtxn) {
+  assert(rtxn->unprepared_);
+  auto cf_map_shared_ptr = WritePreparedTxnDB::GetCFHandleMap();
```
Comment (maysamyabandeh) [inline utilities/transactions/write_unprepared_txn_db.cc]: Can you put a TODO to reduce the duplicate with WritePrepared rollback logic.

File: utilities/transactions/write_unprepared_transaction_test.cc [hunk 6]
```diff
@@ -188,6 +188,54 @@ TEST_P(WriteUnpreparedTransactionTest, ReadYourOwnWrite) {
   delete txn;
 }
 
+TEST_P(WriteUnpreparedTransactionTest, RecoveryRollbackUnprepared) {
```
Comment (maysamyabandeh) [inline utilities/transactions/write_unprepared_transaction_test.cc]: To the test rollback logic further, we can have another test that starts with a non-empty db, and checks that the DB state is the same after recovery.

File: db/write_batch.cc [hunk 7]
```diff
@@ -532,8 +536,9 @@ Status WriteBatch::Iterate(Handler* handler) const {
         break;
       case kTypeBeginUnprepareXID:
         assert(content_flags_.load(std::memory_order_relaxed) &
-               (ContentFlags::DEFERRED | ContentFlags::HAS_BEGIN_PREPARE));
-        handler->MarkBeginPrepare();
+               (ContentFlags::DEFERRED | ContentFlags::HAS_BEGIN_PREPARE |
+                ContentFlags::HAS_BEGIN_UNPREPARE));
```
[the remark this answers is not in the document]
Comment (lth) [reply db/write_batch.cc]: Yes, I agree the assert currently is fairly useless. I think changing it to just `(ContentFlags::DEFERRED | ContentFlags::HAS_BEGIN_UNPREPARE)` would be better.

File: utilities/transactions/write_unprepared_txn_db.cc (hunk 1 above)
Comment (lth) [reply utilities/transactions/write_unprepared_txn_db.cc]: In a subsequent diff, when I set id to '0' when I flush unprepared/prepared batches into DB. For WriteUnpreparedTxn, the transaction should consult unprep_seqs_ instead of just id_.

Looking around, it seems like GetId is mostly used for updating CommitMap/PrepareHeap, and for these operations to be correct here, you would need unprep_seqs_ anyway.

File: utilities/transactions/write_unprepared_txn_db.cc (hunk 2 above)
Comment (lth) [reply utilities/transactions/write_unprepared_txn_db.cc]: I don't think that's really needed. By populating unprep_seqs_, I've already essentially implemented RebuildFromWriteBatch.

I suppose I'm missing WriteBatchInternal::InsertNoop though.

File: utilities/transactions/write_unprepared_txn_db.cc (hunk 3 above)
Comment (lth) [reply utilities/transactions/write_unprepared_txn_db.cc]: I'll document this, but just in case you disagree, this is just to avoid extra work of reconstructing a Transaction.

File: utilities/transactions/write_unprepared_txn_db.cc (hunk 1 above)
Comment (lth) [reply utilities/transactions/write_unprepared_txn_db.cc]: Actually, this is failing transaction_test, since the commit paths still assume that the prepared id (and prepare_batch_cnt) is populated. I'll just set it to the last sequence number.

File: utilities/transactions/write_unprepared_transaction_test.cc [hunk 8]
```diff
@@ -188,6 +188,96 @@ TEST_P(WriteUnpreparedTransactionTest, ReadYourOwnWrite) {
...
+      ASSERT_OK(wup_db->db_impl_->WriteImpl(
+          write_options, &batch, /*callback*/ nullptr, &log_number,
+          /*log ref*/ 0, /* disable memtable */ true, &seq_used,
+          /* prepare_batch_cnt_ */ 1));
+    }
+
+    // Crash and run recovery code paths.
+    wup_db->db_impl_->FlushWAL(true);
+    wup_db->TEST_Crash();
+    ReOpenNoDelete();
```
Comment (maysamyabandeh) [inline utilities/transactions/write_unprepared_transaction_test.cc]: Actually I suggest adding back `wup_db = dynamic_cast<WriteUnpreparedTxnDB*>(db);` after ReOpenNoDelete(). If another engineer later updates the test, they could make use of wup_db assuming it is a valid pointer; it might take hours of debugging before the notice the problem.

File: utilities/transactions/write_unprepared_transaction_test.cc [hunk 9]
```diff
@@ -188,6 +188,96 @@ TEST_P(WriteUnpreparedTransactionTest, ReadYourOwnWrite) {
...
```
Comment (maysamyabandeh) [inline utilities/transactions/write_unprepared_transaction_test.cc]: Same here for updating wup_db pointer.

File: db/db_impl.h [hunk 10]
```diff
@@ -591,6 +589,22 @@ class DBImpl : public DB {
...
       batches_[seq] = {log, batch, batch_cnt};
     }
+
+    ~RecoveredTransaction() {
+      for (auto& it : batches_) {
+        delete it.second.batch_;
+      }
+    }
+
+    void AddBatch(SequenceNumber seq, uint64_t log_number, WriteBatch* batch, size_t batch_cnt, bool unprepared)
```
Comment (maysamyabandeh) [inline db/db_impl.h]: is `make format` run on this patch?

File: db/write_batch.cc [hunk 11]
```diff
@@ -1586,7 +1593,7 @@ class MemTableInserter : public WriteBatch::Handler {
     }
   }
 
-  Status MarkBeginPrepare() override {
+  Status MarkBeginPrepare(bool unprepare) override {
```
Comment (maysamyabandeh) [inline db/write_batch.cc]: since the term "unprepare" is counter-intuitive here, can you add an inline comment to explain it again.

File: db/write_batch.cc [hunk 12]
```diff
@@ -1602,6 +1610,8 @@ class MemTableInserter : public WriteBatch::Handler {
       // we are now iterating through a prepared section
       rebuilding_trx_ = new WriteBatch();
       rebuilding_trx_seq_ = sequence_;
+      unprepared_batch_ = unprepare;
```
[the remark this answers is not in the document]
Comment (maysamyabandeh) [reply db/write_batch.cc]: what if we had two unprepared batches? would not unprepared_batch_ be set true by the first unprepared batch?

File: utilities/transactions/write_unprepared_transaction_test.cc [hunk 13]
```diff
@@ -234,6 +234,48 @@ TEST_P(WriteUnpreparedTransactionTest, RecoveryRollbackUnprepared) {
     ASSERT_FALSE(iter->Valid());
     delete iter;
   }
+
+  // Try unprepared batches with empty database.
```
Comment (maysamyabandeh) [inline utilities/transactions/write_unprepared_transaction_test.cc]: did you mean non-empty?

File: utilities/transactions/write_unprepared_transaction_test.cc [hunk 14]
```diff
@@ -188,6 +188,96 @@ TEST_P(WriteUnpreparedTransactionTest, ReadYourOwnWrite) {
...
+    delete iter;
+  }
+
+  // Try unprepared batches with empty database.
+  for (int num_batches = 0; num_batches < 10; num_batches++) {
+    // Reset database.
+    prepared_trans.clear();
+    ReOpen();
+    wup_db = dynamic_cast<WriteUnpreparedTxnDB*>(db);
+    ASSERT_OK(db->Put(WriteOptions(), "k0", "before value"));
```
Comment (maysamyabandeh) [inline utilities/transactions/write_unprepared_transaction_test.cc]: It would be a stronger test if the initial state of the db has values for k_i where 0<= i < num_batches. In this way we test that the rollback has not messed up the previous value in each batch. Also for k_i lets assign "before value i"

File: utilities/transactions/write_unprepared_transaction_test.cc [hunk 15]
```diff
@@ -188,6 +188,96 @@ TEST_P(WriteUnpreparedTransactionTest, ReadYourOwnWrite) {
...
+  for (int num_batches = 0; num_batches < 10; num_batches++) {
+    // Reset database.
+    prepared_trans.clear();
+    ReOpen();
+    wup_db = dynamic_cast<WriteUnpreparedTxnDB*>(db);
+
+    // Write num_batches unprepared batches into the WAL.
+    for (int i = 0; i < num_batches; i++) {
+      batch.Clear();
+      ASSERT_OK(WriteBatchInternal::InsertNoop(&batch));
```
Comment (maysamyabandeh) [inline utilities/transactions/write_unprepared_transaction_test.cc]: Where do we call MarkBeginPrepare? How is that logic tested?

Comment (maysamyabandeh) [reply utilities/transactions/write_unprepared_transaction_test.cc]: Can you add a TODO to replace this manual construction of the write batch with a call to txn->Write (when that part of the code is done)? It would be hard to read/maintain the low level batch construction code if it is not centralized in one class.

File: db/db_impl.h [hunk 16]
```diff
@@ -600,9 +625,14 @@ class DBImpl : public DB {
 
   void InsertRecoveredTransaction(const uint64_t log, const std::string& name,
                                   WriteBatch* batch, SequenceNumber seq,
-                                  size_t batch_cnt) {
-    recovered_transactions_[name] =
-        new RecoveredTransaction(log, name, batch, seq, batch_cnt);
+                                  size_t batch_cnt, bool unprepared) {
```
Comment (maysamyabandeh) [inline db/db_impl.h]: can we rename it to unprepared_batch to make it easier to distinguish from other unprepared variable?

File: utilities/transactions/write_unprepared_transaction_test.cc (hunk 8 above)
Comment (lth) [reply utilities/transactions/write_unprepared_transaction_test.cc]: OK, I just removed it because of a dead store warning.

File: db/write_batch.cc [hunk 17]
```diff
@@ -1609,6 +1611,8 @@ class MemTableInserter : public WriteBatch::Handler {
       // we are now iterating through a prepared section
       rebuilding_trx_ = new WriteBatch();
       rebuilding_trx_seq_ = sequence_;
+      // We only call MarkBeginPrepare once per batch, and the batch is
```
Comment (maysamyabandeh) [inline db/write_batch.cc]: nit: "the batch is initialized to false" => "unprepared_batch_ is initialized to false"

File: java/rocksjni/writebatchhandlerjnicallback.cc [hunk 18]
```diff
@@ -305,7 +305,7 @@ rocksdb::Status WriteBatchHandlerJniCallback::PutBlobIndexCF(uint32_t column_fam
   }
 }
 
-rocksdb::Status WriteBatchHandlerJniCallback::MarkBeginPrepare() {
+rocksdb::Status WriteBatchHandlerJniCallback::MarkBeginPrepare(bool) {
   m_env->CallVoidMethod(m_jcallback_obj, m_jMarkBeginPrepareMethodId);
```
Comment (maysamyabandeh) [inline java/rocksjni/writebatchhandlerjnicallback.cc]: should it be changed to calling a non-void method?

File: utilities/transactions/write_unprepared_transaction_test.cc [hunk 19]
```diff
@@ -235,20 +240,27 @@ TEST_P(WriteUnpreparedTransactionTest, RecoveryRollbackUnprepared) {
...
     wup_db = dynamic_cast<WriteUnpreparedTxnDB*>(db);
-    ASSERT_OK(db->Put(WriteOptions(), "k0", "before value"));
+    for (int i = 1; i < num_batches; i++) {
+      ASSERT_OK(db->Put(WriteOptions(), "k" + ToString(i),
+                        "before value " + ToString(i)));
+    }
 
     // Write num_batches unprepared batches into the WAL.
-    for (int i = 0; i < num_batches; i++) {
+    for (int i = 1; i < num_batches; i++) {
```
Comment (maysamyabandeh) [inline utilities/transactions/write_unprepared_transaction_test.cc]: nit: why is i changed from 0 to 1? then we writes (num_batches-1) batches instead of num_batches batches.

File: utilities/transactions/write_unprepared_txn_db.cc [hunk 20]
```diff
@@ -15,6 +15,296 @@
...
+            rollback_merge_operands_(rollback_merge_operands) {}
+
+      Status Rollback(uint32_t cf, const Slice& key) {
+        Status s;
+        CFKeys& cf_keys = keys_[cf];
+        if (cf_keys.size() == 0) {  // just inserted
+          auto cmp = comparators_[cf];
+          keys_[cf] = CFKeys(SetComparator(cmp));
+        }
+        auto it = cf_keys.insert(key);
```
Comment (maysamyabandeh) [inline utilities/transactions/write_unprepared_txn_db.cc]: error: declaration shadows a local variable

File: java/rocksjni/writebatchhandlerjnicallback.cc (hunk 18 above)
Comment (lth) [reply java/rocksjni/writebatchhandlerjnicallback.cc]: The void refers to the return value of the method I think.

I looked into this, and we could pass down the bool as an argument, so that it is exposed in the java api. I'm not very familiar with it though, so I'd prefer somebody more knowledgeable do it, if it's needed.

File: utilities/transactions/write_unprepared_transaction_test.cc (hunk 19 above)
Comment (lth) [reply utilities/transactions/write_unprepared_transaction_test.cc]: Yes, should be zero. I only meant to start num_batches at 1.

File: java/rocksjni/writebatchhandlerjnicallback.cc (hunk 18 above)
Comment (maysamyabandeh) [reply java/rocksjni/writebatchhandlerjnicallback.cc]: makes sense. so for the time being can we assert that bool is always false so the java users do not mistakenly use the bool assuming that it would work.
